### ISSUE TITLE
Formatting Issue with Scientific Notation 

### ISSUE THREAD

**Issue by dhilsen** [maths/moodle-qtype_stack#485]
There appears to be an inconsistent behavior with input formatting for scientific notation of the form `a*10^b` for Input Type "Units" compared to "Numerical" (see below for insert stars setting). 

For a numerical input the student can enter either `5*10^2` or `5 * 10^2` and both forms will be accepted. However, for a units input, entering `5 * 10^2 kg` is not accepted as it is interpreted as `5***10^2*kg`. It's interesting to note that by removing the space prior to the unit `5 * 10^2kg` the answer will be accepted. The parser seems to have difficulty when multiplication is initially specified with * and then followed by implied multiplication using a space. Although this mix may seem unusual, it most closely resembles the form of scientific notation which students are accustomed to seeing.

As a side note, Algebraic input will behave in a similar way. For the expression `a*b*c` the following inputs are accepted: `a * b * c` and `a b c`. However, `a * b c` is interpreted as `a***b*c`.

We are using "Insert stars for implied multiplication and for spaces". In physics we find this setting to be the best choice as students naturally tend to insert a space between the number and unit, but will sometimes use no space.

**Comment by aharjula** [comment]
Well that is the old input processing in play and even the new has that behaviour but we can fix it. Basically, the way the insert stars works is dumb and inserts a star into each space it finds that is not inside a `"string"`, it seems obvious that it should check the surroundings of that space and look for existing stars or other operators on either side and then not place the star if any are found. However, the interesting cases are these:
1.  `x +y` => `x+y` or `x*+y`, the first one seems logical
2.  `x + y` => `x+y` or `x*+y`, the first one seems even more logical
3.  `x - y` => `x-y` or `x*-y`, same here
4.  `x  (y+x)` => `x*(y+x)` or `x(x+y)`, here the latter would get processed by other rules to the former anyway.
5. `x / y` => `x/y`

So basically, instead of replacing `\s*` => `*` we want something like `([^\-/\^\*\+=<>])\s*([^\-/\^\*\+=<>])` =>`$1*$2`

**Comment by aharjula** [comment]
Just to let Chris know this is already working in the parser branch was actually already working before I wrote the last comment:
```
= testing "[a * b c, 5 * 10^2 kg, x +y, x / y, 5kg]"=
The test-string was not directly parseable, some corrections required.
The AST is like this before filters:
[a*b*c,5*10^2*kg,x+y,x/y,5*kg]
------------------------------ MP_Root
------------------------------ MP_Statement
------------------------------ MP_List
 -----                         MP_Operation *
 -                             MP_Identifier a
   ---                         MP_Operation * [fixspaces]
   -                           MP_Identifier b
     -                         MP_Identifier c
       ---------               MP_Operation *
       -                       MP_Integer 5
         -------               MP_Operation * [fixspaces]
         ----                  MP_Operation ^
         --                    MP_Integer 10
            -                  MP_Integer 2
              --               MP_Identifier kg
                 ---           MP_Operation +
                 -             MP_Identifier x
                   -           MP_Identifier y
                     ---       MP_Operation /
                     -         MP_Identifier x
                       -       MP_Identifier y
                         ----  MP_Operation * [insertstars]
                         -     MP_Integer 5
                           --  MP_Identifier kg
```
Basically, it only adds stars when need be to get things parsed and most of the spaces in the examples do not require stars. Note that now we know exactly which stars were added and why, which may be handy when rendering error messages or interpreting what went wrong.

**Comment by dhilsen** [comment]
Thank you Matti for addressing this issue. There is one more test case that I would like to suggest.

A space after a plus sign or minus sign that precedes a numerical answer tends to create problems, particularly when units are involved. Entering `- 25*m/s` (space after minus sign) triggers a significant figure error message in our system even though the significant figures are correct. Entering `- 25 m/s` is not accepted as it is interpreted as `-*25*m/s`. Both of these examples are cases in which a mathematical operator appears next to the space and a star should not be inserted.

**Comment by aharjula** [comment]
Ok. That too should be fine in the parser branch, but it does mix two things the significant figures detection and the parser so lets see how things go in the future.
```
= testing "[ - 25*m/s ,  -    25    m    /    s ,  + + 25 m/s]"=
The AST is like this before filters:
[-25*m/s,-25*m/s,++25*m/s]
-------------------------- MP_Root
-------------------------- MP_Statement
-------------------------- MP_List
 -------                   MP_Operation *
 ---                       MP_PrefixOp -
  --                       MP_Integer 25
     ---                   MP_Operation /
     -                     MP_Identifier m
       -                   MP_Identifier s
         -------           MP_Operation * [fixspaces]
         ---               MP_PrefixOp -
          --               MP_Integer 25
             ---           MP_Operation /
             -             MP_Identifier m
               -           MP_Identifier s
                 --------  MP_PrefixOp +
                  -------  MP_Operation * [fixspaces]
                  ---      MP_PrefixOp +
                   --      MP_Integer 25
                      ---  MP_Operation /
                      -    MP_Identifier m
                        -  MP_Identifier s
```

I just noticed an interesting problem with insert stars, if we have something like this:
```
= testing " alpha*kgm/foos"=
The test-string was directly parseable.
The AST is like this before filters:
alpha*kgm/foos
-------------- MP_Root
-------------- MP_Statement
-------------- MP_Operation *
-----          MP_Identifier alpha
      -------- MP_Operation /
      ---      MP_Identifier kgm
          ---- MP_Identifier foos
```
and we apply the single char variables filter in units mode to it we get `alpha*kg*m/f*o*o*s` (if units mode is off we get `alpha*k*g*m/f*o*o*s`) I am starting to wonder which is more correct there, that or splitting the denominator to a group `alpha*kg*m/(f*o*o*s)`. Basically, does the single char variable names feature also imply grouping? And if so does it also apply to every other form of automatic splitting? Mainly an issue in divisions...
